coc/static/update_static.py

```diff
@@ -6,229 +6,1482 @@
-
-    os.unlink("apk.zip")
-    return fingerprint
-
-# Hard-code or fallback
-FINGERPRINT = get_fingerprint()
-BASE_URL = f"https://game-assets.clashofclans.com/{FINGERPRINT}"
-
-def decompress(data):
-    """
-    Decompresses the given bytes 'data' if needed (LZHAM, ZSTD, or LZMA).
-    Returns (decompressed_bytes, compression_details).
-    """
-    if data[0:4] == b"SCLZ":
-        logging.debug("Decompressing using LZHAM ...")
-        import lzham
-
-        dict_size = int.from_bytes(data[4:5], byteorder="big")
-        uncompressed_size = int.from_bytes(data[5:9], byteorder="little")
-
-        decompressed = lzham.decompress(data[9:], uncompressed_size, {"dict_size_log2": dict_size})
-        return decompressed, {
-            "dict_size": dict_size,
-            "uncompressed_size": uncompressed_size,
-        }
+    async def get_fingerprint(self):
+        data = await self.download(self.APK_URL)
 
-    import zstandard
-    if int.from_bytes(data[0:4], byteorder="little") == zstandard.MAGIC_NUMBER:
-        logging.debug("Decompressing using ZSTD ...")
-        decompressed = zstandard.decompress(data)
-        return decompressed, {
-            "dict_size": None,
-            "uncompressed_size": None,
-        }
+        with open("apk.zip", "wb") as f:
+            f.write(data)
+        zf = zipfile.ZipFile("apk.zip")
+        with zf.open('assets/fingerprint.json') as fp:
+            fingerprint = json.loads(fp.read())['sha']
+
+        os.remove("apk.zip")
+        return fingerprint
+
+    def decompress(self, data):
+        """
+        Decompresses the given bytes 'data' if needed (LZHAM, ZSTD, or LZMA).
+        Returns (decompressed_bytes, compression_details).
+        """
+        if data[0:4] == b"SCLZ":
+            logging.debug("Decompressing using LZHAM ...")
+            import lzham
+
+            dict_size = int.from_bytes(data[4:5], byteorder="big")
+            uncompressed_size = int.from_bytes(data[5:9], byteorder="little")
+
+            decompressed = lzham.decompress(data[9:], uncompressed_size, {"dict_size_log2": dict_size})
+            return decompressed, {
+                "dict_size": dict_size,
+                "uncompressed_size": uncompressed_size,
+            }
+
+        if int.from_bytes(data[0:4], byteorder="little") == zstandard.MAGIC_NUMBER:
+            logging.debug("Decompressing using ZSTD ...")
+            decompressed = zstandard.decompress(data)
+            return decompressed, {
+                "dict_size": None,
+                "uncompressed_size": None,
+            }
+
+        # Otherwise, assume LZMA
+        logging.debug("Decompressing using LZMA ...")
+        data = data[0:9] + (b"\x00" * 4) + data[9:]
+        prop = data[0]
+        o_prop = prop
+        if prop > (4 * 5 + 4) * 9 + 8:
+            raise Exception("LZMA properties error")
+        decompressed = lzma.LZMADecompressor().decompress(data)
+        return decompressed, {"lzma_prop": o_prop}
```

Comment:
the dict should contain the keys dict_size and uncompressed_size as the other returns